src/analyses/base.ml

```diff
@@ -769,6 +769,102 @@ struct
+      | BinOp (LOr, arg1, arg2, typ) as exp ->
+        let (let*) = Option.bind in
+        (* split nested LOr Eqs to equality pairs, if possible *)
+        let rec split = function
+          (* copied from above to support pointer equalities with implicit casts inserted *)
+          | BinOp (op, (CastE (t1, e1) as c1), (CastE (t2, e2) as c2), typ) when typeSig t1 = typeSig t2 && (op = Eq || op = Ne) ->
+            let te1 = Cilfacade.typeOf e1 in
+            let te2 = Cilfacade.typeOf e2 in
+            let both_arith_type = isArithmeticType te1 && isArithmeticType te2 in
+            let is_safe = VD.is_safe_cast t1 te1 && VD.is_safe_cast t2 te2 && not both_arith_type in
+            M.tracel "cast" "remove cast on both sides for %a? -> %b\n" d_exp exp is_safe;
+            if is_safe then ( (* we can ignore the casts if the casts can't change the value *)
+              let e1 = if isArithmeticType te1 then c1 else e1 in
+              let e2 = if isArithmeticType te2 then c2 else e2 in
+              Some [(e1, e2)]
+            )
+            else
+              Some [(c1, c2)]
+          | BinOp (Eq, arg1, arg2, _) ->
+            Some [(arg1, arg2)]
+          | BinOp (LOr, arg1, arg2, _) ->
+            let* s1 = split arg1 in
+            let* s2 = split arg2 in
+            Some (s1 @ s2)
+          | _ ->
+            None
+        in
+        (* find common exp from all equality pairs and list of other sides, if possible *)
+        let find_common = function
+          | [] -> assert false
+          | (e1, e2) :: eqs ->
+            let eqs_for_all_mem e = List.for_all (fun (e1, e2) -> CilType.Exp.(equal e1 e || equal e2 e)) eqs in
+            let eqs_map_remove e = List.map (fun (e1, e2) -> if CilType.Exp.equal e1 e then e2 else e1) eqs in
+            if eqs_for_all_mem e1 then
+              Some (e1, e2 :: eqs_map_remove e1)
+            else if eqs_for_all_mem e2 then
+              Some (e2, e1 :: eqs_map_remove e2)
+            else
+              None
+        in
+        let eqs_value =
+          let* eqs = split exp in
+          let* (e, es) = find_common eqs in
+          let v = eval_rv a gs st e in (* value of common exp *)
+          let vs = List.map (eval_rv a gs st) es in (* values of other sides *)
+          let ik = Cilfacade.get_ikind typ in
+          match v with
+          | `Address a ->
+            (* get definite addrs from vs *)
+            let rec to_definite_ad = function
+              | [] -> AD.empty ()
+              | `Address a :: vs when AD.is_definite a ->
+                AD.union a (to_definite_ad vs)
+              | _ :: vs ->
+                to_definite_ad vs
+            in
+            let definite_ad = to_definite_ad vs in
+            if AD.leq a definite_ad then (* other sides cover common address *)
+              Some (`Int (ID.of_bool ik true))
+            else if AD.is_empty (AD.meet a definite_ad) then
+              Some (`Int (ID.of_bool ik false))
```

Comment:
Is this not unsound for `LOr`, in combination with discarding non-definite address in line 826?
Say we have the following expression, with `p`,`q` being pointers and `w`, `x`, `y`,`z` being variables of the corresponding type.
```C
p==&x|| p==q
```
Assume that possible values are: `p = {&w, &y, &z}` and `q = {&y, &z}`. Then the expression may evaluate to `true`. The meet of `a`  (==  {&w, &y, &z}) and `definite_ad = {&x}` would be empty, however; thus yielding a definite `false`. 

Using the join over `vs` in place of `definite_ad` should work here.

Reply:
You're right! I added that disjoint check at the last moment without thinking it through. For simplicity I just removed both for now since refuting those witness invariants isn't the current goal. In the ints case it shouldn't be then based on just inclusion sets anyway, but could be more general, so I'll leave it for now.